### PR TITLE
Add Match to Option

### PR DESCRIPTION
**Pull request by zaudtke**
Match added to Option and tests added

### CODE REVIEW

File: Optionally/Options.cs [hunk 1]
```diff
@@ -34,6 +34,15 @@ public interface IOption<T>
         /// <param name="filter">Predicate used to test Option</param>
         /// <returns>If Option is Some and the value fulfills the filter, then Some. Otherwise None</returns>
         IOption<T> Where(Func<T, bool> filter);
+
+        /// <summary>
+        /// Performs a Fuunction on the current Options
```
Comment (cameronpresley) [inline Optionally/Options.cs]: Small typo, believe you meant _Function_ not _Fuunction_

File: Optionally.Tests/OptionTests/WhenMatching.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            int expected = 4;
+            var some = Option.Some(4);
+
+            var value = some.Match(
+                ifNone: () => 0,
+                ifSome: x => x);
+
+            Assert.AreEqual(expected, value);
+        }
+
```
Comment (cameronpresley) [inline Optionally.Tests/OptionTests/WhenMatching.cs:25]: Total nitpick, but we should think about adding another test that made sure that the function was actually called. At this point, all we know is that a value was returned, but we don't know if the value was hard-coded. A way we can achieve that style of testing looks a lot like below
```
bool someWasCalled = false;
var some = Option.Some(4);

var value = some.Match(
     ifNone: () => 0,
     ifSome: x => {someWasCalled=true; return x;});

Assert.That(someWasCalled);
```

**Review by cameronpresley — CHANGES_REQUESTED** [review]
This looks awesome, but there are two improvement requests. One is a typo and the other is a possible test case that was missed.

File: Optionally/Options.cs [hunk 3]
```diff
@@ -34,6 +34,15 @@ public interface IOption<T>
         /// <param name="filter">Predicate used to test Option</param>
         /// <returns>If Option is Some and the value fulfills the filter, then Some. Otherwise None</returns>
         IOption<T> Where(Func<T, bool> filter);
+
+        /// <summary>
+        /// Performs a Function on the current Option
+        /// </summary>
+        /// <typeparam name="U">Type to convert to</typeparam>
+        /// <param name="ifNone">Function to call if Option is None</param>
+        /// <param name="ifSome">Function to call if Option is None</param>
```
Comment (cameronpresley) [inline Optionally/Options.cs]: For ifSome, probably should read _Option is Some_ instead of _Option is None_

**Review by cameronpresley — CHANGES_REQUESTED** [review]

File: Optionally/Options.cs (hunk 1 above)
Comment (zaudtke) [reply Optionally/Options.cs]: Changed and commited

File: Optionally/Options.cs (hunk 3 above)
Comment (zaudtke) [reply Optionally/Options.cs]: Changed and committed

**Review by cameronpresley — APPROVED** [review]

File: Optionally.Tests/OptionTests/WhenMatching.cs (hunk 2 above)
Comment (zaudtke) [reply Optionally.Tests/OptionTests/WhenMatching.cs:25]: Are you wanting both, or use this to replace the existing?

Comment (zaudtke) [reply Optionally.Tests/OptionTests/WhenMatching.cs:25]: Tests now check that function ran, instead of return value.

**Pull request merged**